### PR TITLE
perf: virtual-core rewrite for mount/measure-storm, plus iOS Safari handling and scroll restoration

### PR DESCRIPTION
**Pull request by tannerlinsley**
## Summary

A perf-focused release covering several themes. Release blog post at [tanstack.com#934](https://github.com/TanStack/tanstack.com/pull/934).

**Six changesets:**

| Bump | Package | Theme |
|---|---|---|
| minor | `@tanstack/virtual-core` | Lazy `VirtualItem` materialization (typed-array + Proxy) for the lanes===1 fast path, plus eight smaller audit hotfixes (Map clone, setOptions deopt, Math.min spread, defaultRangeExtractor allocation, elementsCache leak, etc.) |
| minor | `@tanstack/virtual-core` | iOS Safari momentum-scroll handling (touch event distinction, subpixel reconciliation, elastic-overscroll clamp) |
| minor | `@tanstack/virtual-core` | Skip scroll-position adjustment on backward scroll by default (closes the largest single complaint cluster) |
| minor | `@tanstack/virtual-core` | New `takeSnapshot()` public method for scroll-restoration round-trips |
| patch | `@tanstack/virtual-core` | `scrollToIndex(N, { behavior: 'smooth' })` keeps smooth while still > viewport away from target instead of snapping to 'auto' on first retarget |
| patch | `@tanstack/react-virtual` | Use a number counter for the force-rerender pattern, dropping the `{}` allocation per scroll dispatch |

## Headline numbers

| Metric | Before | After |
|---|---|---|
| Cold mount @ 100k items (real React) | 6.1 ms | 4.5 ms |
| Cold mount @ 100k items (synthetic) | 2.5 ms | 0.54 ms |
| Cold mount @ 500k items (synthetic) | 14 ms | 2.7 ms |
| `resizeItem` storm on 10k items | 1.9 s | 1.3 ms |
| `setOptions` × 10,000 (per render) | 14.4 ms | 1.3 ms |
| `scrollToIndex` landing accuracy (dynamic 10k) | within 1 px | 0.0 px |
| iOS Safari momentum scroll | broken | works |
| Backward-scroll jank with dynamic items | recurring | gone by default |

The 1382x `resizeItem` win is real and embarrassing in equal measure (Map clone on every resize purely to invalidate a memo dep).

## Bundle size

| Build | origin/main | This PR | Delta |
|---|---:|---:|---:|
| Consumer-minified gzip (esbuild prod) | 5.22 kB | 6.11 kB | +890 B (+17%) |

About 430 B for the lazy-materialization machinery, 370 B for iOS handling, and 90 B for the rest of the fixes combined.

## Behavior changes

Three default changes consumers should know about:

1. Backward-scroll no longer writes `scrollTop` on above-viewport resize by default. Users who relied on the old behavior can supply `shouldAdjustScrollPositionOnItemSizeChange`.
2. iOS Safari adjustments are deferred until scroll settles. Invisible to most users; fixes recurring bug reports.
3. `setOptions` no longer mutates the caller's options object (hidden contract violation; unlikely anyone was relying on the mutation).

## What's new

- `virtualizer.takeSnapshot()` — returns currently-measured items as plain `VirtualItem` objects, pairs with `initialMeasurementsCache` for scroll restoration after navigation. Documented in `docs/api/virtualizer.md`.
- `initialMeasurementsCache` is now documented (was previously undocumented even though the option already existed).

## New tests

91 unit tests across `virtual-core` and `react-virtual`, up from 13 originally. New coverage for: lazy fast-path edge cases (9), iOS deferral (7), subpixel reconciliation (3), elastic-overscroll clamp (3), backward-scroll skip (3), takeSnapshot round-trip (2), several smaller audit-driven regression tests.

## Reproducible benchmark suite

Added `benchmarks/` — a Vite app + Playwright runner that drives the same scenarios across `@tanstack/react-virtual` and the other major virtualization libraries and reports medians across runs. Documented in `benchmarks/README.md`. Run with `cd benchmarks && pnpm bench`. Useful as a regression guard for future perf work.

## Test plan

- [x] `pnpm test:lib` (91/91 passing)
- [x] `pnpm test:types`
- [x] `pnpm test:eslint`
- [x] `pnpm test:build`
- [x] `pnpm test:knip`
- [x] `pnpm test:sherif`
- [x] `pnpm test:docs` (link check)
- [x] `pnpm changeset` schema validates
- [x] React-virtual integration tests (6/6 passing)
- [x] Cross-library benchmark suite runs end-to-end
- [ ] Optional: rerun `pnpm bench` on reviewer's machine to confirm numbers
- [ ] Optional: `pnpm changeset:version` preview before merging

Pre-existing `lit-virtual:build` and `:test:e2e` failures on main are unrelated to this PR.

## Summary by CodeRabbit

* **New Features**
  * Added takeSnapshot() and initialMeasurementsCache for snapshot/restore.
  * Default iOS Safari momentum-scroll protections enabled.
  * Skip scroll-position adjustment when scrolling backward by default.

* **Bug Fixes**
  * Smooth scroll preserves smooth behavior on dynamic-height lists; final landing accuracy improved.
  * Prevented upward-scroll jank from above-viewport resizes.

* **Performance**
  * Lower allocations and faster cold-mounts via lazy measurements and hot-path optimizations.
  * Reduced per-dispatch allocations in React integration.

* **Benchmarks**
  * New reproducible browser benchmark harness, runner, datasets and sample results; added benchmarks workspace.

* **Documentation**
  * Updated Virtualizer API docs for new options and methods.

* **Tests**
  * Expanded coverage and benchmarks for scroll, measurement, iOS, and reconciliation behaviors.

[Review Change Stack](https://app.coderabbit.ai/change-stack/TanStack/virtual/pull/1168?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: .claude/scheduled_tasks.lock [hunk 1]
```diff
@@ -0,0 +1 @@
+{"sessionId":"e402bf71-ca74-4aa5-856c-da0c2053caab","pid":78596,"procStart":"Sat May 16 20:13:35 2026","acquiredAt":1779000018499}
```
Comment (SeanCassiere) [inline .claude/scheduled_tasks.lock:1]: Kill?

File: packages/virtual-core/src/index.ts [hunk 2]
```diff
@@ -1,6 +1,29 @@
...
+let _isIOSResult: boolean | undefined
+const isIOSWebKit = (): boolean => {
+  if (_isIOSResult !== undefined) return _isIOSResult
+  if (typeof navigator === 'undefined') return (_isIOSResult = false)
+  if (/iP(hone|od|ad)/.test(navigator.userAgent)) return (_isIOSResult = true)
+  // iPadOS 13+ reports as MacIntel; touch-points distinguishes it from desktop.
+  const mtp = (navigator as Navigator & { maxTouchPoints?: number })
+    .maxTouchPoints
+  return (_isIOSResult =
+    navigator.platform === 'MacIntel' && mtp !== undefined && mtp > 0)
```
Comment (piecyk) [inline packages/virtual-core/src/index.ts:17]: Looks like `navigator.platform` is [deprecated](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/platform). It still works today, but could be replaced with `navigator.userAgentData?.platform` with a fallback. 

```suggestion
  const nav = navigator as Navigator & {
    userAgentData?: {
      platform?: string
    }
    maxTouchPoints?: number
  }

  if (/\b(iPhone|iPod|iPad)\b/.test(nav.userAgent)) {
    return (_isIOSResult = true)
  }

  // iPadOS 13+ may report as macOS/MacIntel; touch-points distinguish it from desktop.
  // Use userAgentData.platform when available, otherwise fall back to navigator.platform.
  const platform = nav.userAgentData?.platform ?? nav.platform
  const maxTouchPoints = nav.maxTouchPoints ?? 0

  return (_isIOSResult =
    (platform === 'macOS' || platform === 'MacIntel') && maxTouchPoints > 0)```

**Review by piecyk — APPROVED** [review]
LGTM 👍 Impressive rewrite. The core perf wins are well-motivated and clearly implemented, let's 🚢 🇮🇹

File: packages/virtual-core/src/index.ts (hunk 2 above)
Comment (KevinVandy) [reply packages/virtual-core/src/index.ts:17]: I had a couple iterations of safely accessing navigator.platform in all environments safely (like even cloudflare ssr stuff) and I ended up at this util in hotkeys https://github.com/TanStack/hotkeys/blob/404666b65dbec31d47da1be4c2a2210e7e53a615/packages/hotkeys/src/constants.ts#L26
